gpio/__init__.py

```diff
@@ -0,0 +1,268 @@
+            return f.read().strip()
+
+    def set_direction(self, mode):
+        '''Set the direction of pin
+
+        Args:
+            mode (str): use either gpio.OUT or gpio.IN
+        '''
+        if mode not in (IN, OUT, LOW, HIGH):
+            raise ValueError("Unsupported pin mode {}".format(mode))
+
+        with open(os.path.join(self.root, 'direction'), FMODE) as f:
+            f.write(str(mode))
+            f.flush()
+
+    def set_active_low(self, active_low):
+        '''Set the polarity of pin
+
+        Args:
+            mode (bool): True = active low / False = active high
+        '''
+        if not isinstance(active_low, bool):
+            raise ValueError("active_low must be True or False")
+
+        with open(os.path.join(self.root, 'active_low'), FMODE) as f:
+            f.write('1' if active_low else '0')
+            f.flush()
+
+    def read(self):
+        '''Read pin value
+
+        Returns:
+            int: gpio.HIGH or gpio.LOW
+        '''
+        self.value.seek(0)
+        value = self.value.read()
+        try:
+            # Python > 3 - bytes
+            # Subtracting 48 converts an ASCII "0" or "1" to an int
+            # ord("0") == 48
+            return value[0] - 48
+        except TypeError:
+            # Python 2.x - str
+            return int(value)
+
+    def write(self, value):
+        '''Write pin value
+
+        Args:
+            value (bool): use either gpio.HIGH or gpio.LOW
+        '''
+        # Convert any truthy value explicitly to HIGH and vice versa
+        # this is about 3x faster than int(bool(value))
+        value = HIGH if value else LOW
+        # write as bytes, about 3x faster than string IO
+        self.value.write(b'1' if value else b'0')
+        # state.value.write(str(value).encode())  # Slow alternate for Python 2
+
+    def cleanup(self):
+        '''Clean up pin
+
+        Unexports the pin and deletes it from the open list.
+
+        '''
+        # Note: I have not put "cleanup" into the __del__ method since it's not
+        # always desireable to unexport pins at program exit.
+        # Additionally "open" can be deleted *before* the GPIOPin instance.
+        self.value.close()
+
+        if os.path.exists(self.root):
+            with _export_lock:
+                with open(GPIO_UNEXPORT, FMODE) as f:
+                    f.write(str(self.pin))
+                    f.flush()
+
+        del _open_pins[self.pin]
+
+
+def cleanup(pin=None, assert_exists=False):
+    """Cleanup the pin by closing and unexporting it.
+
+    Args:
+        pin (int, optional): either the pin to clean up or None (default).
+            If None, clean up all pins.
+        assert_exists: if True, raise a ValueError if the pin was not
+            setup. Otherwise, this function is a NOOP.
+    """
+    # Note: since "pin" is a kwarg in this function, it has not been renamed it to "pins" above
+    pins = pin
+
+    if pins is None:

```

Comment:
replace these lines with:

```
    if pins is None:
        pins = _open_pins.keys()

    if not isinstance(pins, Iterable)
        pins = [pins]
```

Then add `from collections import Iterable`

Reply:
Bold! Support for generators?

Reply:
The conversion from the generator `_open_pins.keys()` to a fixed list here is required, since the dictionary is modified by the loop that consumes these values. Otherwise +1 for generator support for passed in values.

Reply:
good call.